File: ink_engine/ink_engine/core/llm/messages.py
"""自写消息数据类与工具调用增量累积。

消息 = 单一 Message 数据类（role/content/tool_calls/tool_call_id/reasoning），
工厂函数按角色构造；to_openai_dict 序列化为 OpenAI 兼容请求负载。
工具调用以增量形式从流式 chunk 累积（accumulate_tool_calls 按 index 合并），
与 OpenAI 兼容协议的工具调用分片语义一一对应。
"""
from __future__ import annotations

import json
import uuid
from collections.abc import Iterable
from dataclasses import dataclass
from typing import Any

from ink_engine.core.llm.errors import LLMConfigError, LLMFormatError

_ROLES = frozenset({"system", "user", "assistant", "tool"})

# 历史/外部消息形态的角色别名（human/ai 命名 → 引擎规范角色）
_ROLE_ALIASES = {"human": "user", "ai": "assistant"}

# 支持的附件类型（多模态消息的附件类别；kind 枚举化防魔法字符串）
_ATTACHMENT_KINDS = frozenset({"image", "video", "document"})

# 附件序列化段类型映射：OpenAI 兼容请求只原生定义 image_url 段；
# video/document 按同族收敛形态（<kind>_url，与 Qwen/GLM 等多模态
# 端点惯例一致），适配器按后端支持裁剪（附件由宿主显式声明，
# 声明方负责后端兼容性）。
_ATTACHMENT_SEGMENT_TYPES = {
    "image": "image_url",
    "video": "video_url",
    "document": "document_url",
}


@dataclass(frozen=True, slots=True)
class Attachment:
    """多模态附件元数据（附加在 user 消息上的附件段）。

    Attributes:
        kind: 附件类型（image/video/document）。
        url: 附件远端地址（http/https/data URI；适配器可直接请求的形态）。
        path: 附件本地路径（宿主/本地端点可解析；url 缺省时作为回退引用）。
        mime_type: MIME 类型（如 image/png）。
        alt: 替代文本（辅助功能与降级展示）。
        width/height: 像素尺寸（image 类型）。
        duration: 时长（秒，video 类型）。
        name: 原始文件名（展示/诊断）。
    """

    kind: str = "image"
    url: str | None = None
    path: str | None = None
    mime_type: str | None = None
    alt: str | None = None
    width: int | None = None
    height: int | None = None
    duration: float | None = None
    name: str | None = None

    def __post_init__(self) -> None:
        if self.kind not in _ATTACHMENT_KINDS:
            raise LLMConfigError(f"非法附件类型: {self.kind!r}")
        if not self.url and not self.path:
            raise LLMConfigError("附件必须携带 url 或 path（引用缺失无法发送）")

    @property
    def ref(self) -> str:
        """引用值：url 优先，缺省回落 path（本地端点场景）。"""
        return self.url or self.path or ""

    def to_openai_segment(self) -> dict[str, Any]:
        """序列化为 OpenAI 兼容多模态内容段（{type: <segment>, <segment>: {url}}）。"""
        segment = _ATTACHMENT_SEGMENT_TYPES[self.kind]
        return {"type": segment, segment: {"url": self.ref}}

    def to_dict(self) -> dict[str, Any]:
        return {
            "kind": self.kind,
            "url": self.url,
            "path": self.path,
            "mime_type": self.mime_type,
            "alt": self.alt,
            "width": self.width,
            "height": self.height,
            "duration": self.duration,
            "name": self.name,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> Attachment:
        fields = (
            "kind",
            "url",
            "path",
            "mime_type",
            "alt",
            "width",
            "height",
            "duration",
            "name",
        )
        return cls(**{f: data.get(f) for f in fields})


@dataclass(frozen=True, slots=True)
class ToolCallDelta:
    """工具调用增量帧（按 index 归属一次调用，流式累积用）。"""

    index: int
    id: str | None = None
    name: str | None = None
    arguments_delta: str | None = None


@dataclass(slots=True)
class ToolCall:
    """一次工具调用（增量累积过程中可变——accumulate_tool_calls 就地补全）。"""

    id: str
    name: str
    arguments: str = ""

    @property
    def parsed_arguments(self) -> dict[str, Any]:
        """容错解析 arguments JSON（未完成/非法时返回空 dict，不抛错）。"""
        return self.parse_arguments()

    def parse_arguments(self, *, strict: bool = False) -> dict[str, Any]:
        """解析 arguments JSON。

        容错（默认）：未完成/非法返回空 dict——流式累积中的截断碎片是
        常态，容忍解析不抛错；调用方在**执行前**须用 strict=True 校验
        完整参数（防"参数被截断却以空参数执行"的静默降级）。
        strict=True：非法/非对象 JSON 抛 LLMFormatError，调用方显式拒绝。
        """
        if not self.arguments.strip():
            if strict:
                raise LLMFormatError("工具调用参数为空或未完成")
            return {}
        try:
            value = json.loads(self.arguments)
        except json.JSONDecodeError:
            if strict:
                raise LLMFormatError("工具调用参数非法（非完整 JSON）") from None
            return {}
        if not isinstance(value, dict):
            if strict:
                raise LLMFormatError("工具调用参数须为 JSON 对象")
            return {}
        return value


@dataclass(slots=True)
class Message:
    """聊天消息（system/user/assistant/tool 四角色）。

    Args:
        id: 消息唯一 id（add_messages 按 id 去重/RemoveMessage 删除；None = 追加不去重，
            构造时未显式提供则自动生成）。
        role: 角色（system/user/assistant/tool）。
        content: 文本内容（assistant 可为空——纯工具调用消息）。
        tool_call_id: tool 角色必填，回指 assistant 的工具调用 id。
        tool_calls: assistant 角色的工具调用列表（历史消息回传用）。
        reasoning: assistant 的推理文本（历史消息回传时可选携带，
            当前轮推理以流式 reasoning_token 增量产出，不入此字段）。
        attachments: user 角色的多模态附件（默认空元组；携带时
            to_openai_dict 以内容段数组形态序列化）。
    """

    role: str
    content: str = ""
    tool_call_id: str | None = None
    tool_calls: list[ToolCall] | None = None
    reasoning: str | None = None
    id: str | None = None
    attachments: tuple[Attachment, ...] = ()

    def __post_init__(self) -> None:
        if self.role not in _ROLES:
            raise LLMConfigError(f"非法消息角色: {self.role!r}")
        if self.role == "tool" and not self.tool_call_id:
            raise LLMConfigError("tool 角色消息必须携带 tool_call_id")
        if self.id is None:
            self.id = uuid.uuid4().hex
        # 归一为不可变元组（list/dict 形态的宽容入参 → 规范形态），
        # 保持消息可哈希引用语义（tuple 字段冻结，序列化确定性）
        self.attachments = tuple(
            a if isinstance(a, Attachment) else Attachment.from_dict(a)
            for a in self.attachments
        )

    def to_openai_dict(self) -> dict[str, Any]:
        """序列化为 OpenAI 兼容请求负载。

        user 消息携带附件时 content 序列化为多模态内容段数组
        （[{"type": "text", ...}, {"type": "image_url", ...}, ...]）；
        无附件时输出形态与既往逐字段一致（回归零影响）。
        """
        if self.role == "tool":
            return {"role": "tool", "content": self.content, "tool_call_id": self.tool_call_id}
        if self.role == "user" and self.attachments:
            content: list[dict[str, Any]] = []
            if self.content:
                content.append({"type": "text", "text": self.content})
            content.extend(a.to_openai_segment() for a in self.attachments)
            return {"role": "user", "content": content}
        payload: dict[str, Any] = {"role": self.role, "content": self.content}
        if self.role == "assistant" and self.tool_calls:
            payload["tool_calls"] = [
                {
                    "id": tc.id,
                    "type": "function",
                    "function": {"name": tc.name, "arguments": tc.arguments},
                }
                for tc in self.tool_calls
            ]
        return payload

    def to_dict(self) -> dict[str, Any]:
        return {
            "role": self.role,
            "content": self.content,
            "tool_call_id": self.tool_call_id,
            "tool_calls": (
                [
                    {"id": tc.id, "name": tc.name, "arguments": tc.arguments}
                    for tc in self.tool_calls
                ]
                if self.tool_calls
                else None
            ),
            "reasoning": self.reasoning,
            "id": self.id,
            "attachments": (
                [a.to_dict() for a in self.attachments] if self.attachments else None
            ),
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> Message:
        tool_calls = data.get("tool_calls")
        attachments = data.get("attachments")
        return cls(
            role=data["role"],
            content=data.get("content", ""),
            tool_call_id=data.get("tool_call_id"),
            tool_calls=(
                [ToolCall(**tc) for tc in tool_calls] if tool_calls else None
            ),
            reasoning=data.get("reasoning"),
            id=data.get("id"),
            attachments=(
                tuple(Attachment.from_dict(a) for a in attachments)
                if attachments
                else ()
            ),
        )


def system(text: str) -> Message:
    return Message(role="system", content=text)


def user(
    text: str,
    *,
    attachments: Iterable[Attachment] | None = None,
) -> Message:
    return Message(
        role="user",
        content=text,
        attachments=tuple(attachments or ()),
    )


def assistant(text: str = "", *, tool_calls: list[ToolCall] | None = None, reasoning: str | None = None) -> Message:
    return Message(role="assistant", content=text, tool_calls=tool_calls, reasoning=reasoning)


def tool_result(content: str, tool_call_id: str) -> Message:
    return Message(role="tool", content=content, tool_call_id=tool_call_id)


def message_role(msg: Any) -> str:
    """任意消息形态的角色归一（system/user/assistant/tool）。

    消息在引擎执行期统一为 Message，但状态通道里可能混入宿主注入的 dict
    形态（初始输入、删除标记）或历史遗留的鸭子类型消息类——上下文投影、
    窗口裁剪等原语需要一个总函数来判角色，不能假定单一形态。

    识别顺序：

    1. ``role`` 属性（引擎 Message，规范角色直接透传）；
    2. dict 的 ``role`` / ``type`` 键（``type`` 是宿主注入形态的常用键）；
    3. 类名兜底（去掉 ``Message`` 后缀，如 ``AIMessage`` → assistant）。

    human/ai 别名统一归一为 user/assistant；无法识别时返回小写类名或空串，
    调用方按「非已知角色」处理即可（不抛错，防迁移期偶发形态崩）。
    """
    role = getattr(msg, "role", None)
    if role is not None:
        role = str(role)
        return _ROLE_ALIASES.get(role, role)
    if isinstance(msg, dict):
        raw = str(msg.get("role") or msg.get("type") or "")
        return _ROLE_ALIASES.get(raw, raw)
    name = type(msg).__name__.lower()
    # 后缀剥离后为空（类名恰为 "Message"）时退回原类名，避免归一成空串
    stem = name.removesuffix("message") or name
    return _ROLE_ALIASES.get(stem, stem)


def accumulate_tool_calls(deltas: Iterable[ToolCallDelta]) -> list[ToolCall]:
    """按 index 合并流式工具调用增量。

    OpenAI 兼容协议把一次工具调用拆成多个 delta（id/name 首次出现，
    arguments 逐帧拼接）；本函数按 index 累积为完整 ToolCall 列表
    （保持首次出现顺序，index 乱序时仍按首见序输出）。
    """
    by_index: dict[int, ToolCall] = {}
    order: list[int] = []
    for delta in deltas:
        tc = by_index.get(delta.index)
        if tc is None:
            tc = by_index[delta.index] = ToolCall(id=delta.id or "", name=delta.name or "", arguments="")
            order.append(delta.index)
        if delta.id:
            tc.id = delta.id
        if delta.name:
            tc.name = delta.name
        if delta.arguments_delta:
            tc.arguments += delta.arguments_delta
    return [by_index[i] for i in order]


__all__ = [
    "Attachment",
    "Message",
    "ToolCall",
    "ToolCallDelta",
    "accumulate_tool_calls",
    "assistant",
    "message_role",
    "system",
    "tool_result",
    "user",
]

File: ink-ts/engine/src/kernel/llm/messages.ts
/**
 * 聊天消息与工具调用增量累积（Message + 工厂函数 + accumulate_tool_calls + message_role）。
 *
 * 数据形态类 Attachment / ToolCall / ToolCallDelta 拆入 _shapes.ts 以遵守单文件
 * 行数上限；本模块聚焦 Message 主对象、角色工厂、序列化与角色归一。
 *
 * message_role 作为消息形态角色归一的总入口（覆盖 dict/鸭子类型/类名兜底），
 * 供上下文投影、窗口裁剪等原语复用。
 */

import { LLMConfigError } from './errors.js';
import {
  Attachment,
  ToolCall,
  ToolCallDelta,
  type Json,
} from './_shapes.js';
import { ROLES, ROLE_ALIASES } from './_types.js';

export { Attachment, ToolCall, ToolCallDelta, type Json } from './_shapes.js';

/** uuid 注入面：缺省固定 hex，保证纯函数可复现。 */
export interface UuidOptions {
  /** 十六进制 uuid 源；未注入时按确定值 0...0 × 32 位。 */
  uuid?: () => string;
}

const DEFAULT_UUID = (): string => '00000000000000000000000000000000';

/** 聊天消息（system/user/assistant/tool 四角色）。 */
export class Message {
  readonly role: string;
  readonly content: string;
  readonly tool_call_id: string | null;
  readonly tool_calls: ToolCall[] | null;
  readonly reasoning: string | null;
  readonly id: string;
  readonly attachments: readonly Attachment[];
  readonly name: string | null;

  constructor(
    role: string,
    content: string = '',
    tool_call_id: string | null = null,
    tool_calls: ToolCall[] | null = null,
    reasoning: string | null = null,
    id: string | null = null,
    attachments: readonly Attachment[] = [],
    name: string | null = null,
    options: UuidOptions = {},
  ) {
    if (!ROLES.has(role)) {
      throw new LLMConfigError(`非法消息角色: ${JSON.stringify(role)}`);
    }
    if (role === 'tool' && !tool_call_id) {
      throw new LLMConfigError('tool 角色消息必须携带 tool_call_id');
    }
    const normAttachments: Attachment[] = [];
    for (const a of attachments) {
      normAttachments.push(a instanceof Attachment ? a : Attachment.from_dict(a as Record<string, Json>));
    }
    const normToolCalls: ToolCall[] | null = tool_calls
      ? tool_calls.map((tc) => (tc instanceof ToolCall ? tc : new ToolCall(tc as { id: string; name: string })))
      : null;
    this.role = role;
    this.content = content;
    this.tool_call_id = tool_call_id;
    this.tool_calls = normToolCalls;
    this.reasoning = reasoning;
    this.id = id ?? (options.uuid ?? DEFAULT_UUID)();
    this.attachments = normAttachments;
    this.name = name;
  }

  /** 序列化为 OpenAI 兼容请求负载（user 多模态附件展开为内容段数组）。 */
  to_openai_dict(): Record<string, Json> {
    if (this.role === 'tool') {
      return { role: 'tool', content: this.content, tool_call_id: this.tool_call_id };
    }
    if (this.role === 'user' && this.attachments.length > 0) {
      const content: { type: string; [key: string]: Json }[] = [];
      if (this.content) content.push({ type: 'text', text: this.content });
      for (const a of this.attachments) content.push(a.to_openai_segment());
      const payload: Record<string, Json> = { role: 'user', content };
      if (this.name) payload['name'] = this.name;
      return payload;
    }
    const payload: Record<string, Json> = { role: this.role, content: this.content };
    if (this.name && (this.role === 'user' || this.role === 'assistant')) {
      payload['name'] = this.name;
    }
    if (this.role === 'assistant' && this.tool_calls) {
      payload['tool_calls'] = this.tool_calls.map((tc) => ({
        id: tc.id,
        type: 'function',
        function: { name: tc.name, arguments: tc.arguments },
      }));
    }
    return payload;
  }

  to_dict(): Record<string, Json> {
    return {
      role: this.role,
      content: this.content,
      tool_call_id: this.tool_call_id,
      tool_calls: this.tool_calls ? this.tool_calls.map((tc) => tc.to_dict()) : null,
      reasoning: this.reasoning,
      id: this.id,
      attachments: this.attachments.length > 0 ? this.attachments.map((a) => a.to_dict()) : null,
      name: this.name,
    };
  }

  static from_dict(data: Record<string, Json>): Message {
    const tcRaw = data['tool_calls'];
    const attRaw = data['attachments'];
    const tool_calls: ToolCall[] | null = Array.isArray(tcRaw)
      ? tcRaw.map((tc) => ToolCall.from_dict(tc as Record<string, Json>))
      : null;
    const attachments: Attachment[] = Array.isArray(attRaw)
      ? attRaw.map((a) => Attachment.from_dict(a as Record<string, Json>))
      : [];
    return new Message(
      data['role'] as string,
      (data['content'] as string | undefined) ?? '',
      (data['tool_call_id'] as string | null | undefined) ?? null,
      tool_calls,
      (data['reasoning'] as string | null | undefined) ?? null,
      (data['id'] as string | null | undefined) ?? null,
      attachments,
      (data['name'] as string | null | undefined) ?? null,
    );
  }
}

export function system(text: string, options: UuidOptions = {}): Message {
  return new Message('system', text, null, null, null, null, [], null, options);
}

export function user(
  text: string,
  options: { attachments?: readonly Attachment[]; name?: string | null; uuid?: () => string } = {},
): Message {
  const { attachments, name, uuid } = options;
  return new Message('user', text, null, null, null, null, attachments ?? [], name ?? null, { uuid });
}

export function assistant(
  text: string = '',
  options: {
    tool_calls?: ToolCall[] | null;
    reasoning?: string | null;
    name?: string | null;
    uuid?: () => string;
  } = {},
): Message {
  const { tool_calls, reasoning, name, uuid } = options;
  return new Message('assistant', text, null, tool_calls ?? null, reasoning ?? null, null, [], name ?? null, { uuid });
}

export function tool_result(content: string, tool_call_id: string, options: UuidOptions = {}): Message {
  return new Message('tool', content, tool_call_id, null, null, null, [], null, options);
}

/**
 * 任意消息形态的角色归一（system/user/assistant/tool）。
 *
 * 识别顺序：role 属性（引擎 Message）→ dict 的 role/type 键 → 类名兜底。
 * human/ai 别名统一归一为 user/assistant；无法识别时返回空串或小写类名，
 * 调用方按「非已知角色」处理即可（不抛错，防迁移期偶发形态崩）。
 *
 * 纯对象字面量（构造来源为 Object/无自定义类名）不触发类名兜底——避免
 * `{}` 被误归为 'object'；类名兜底仅在自定义类实例上生效。
 */
export function message_role(msg: unknown): string {
  const isObj = msg !== null && typeof msg === 'object' && !Array.isArray(msg);
  if (isObj) {
    const obj = msg as Record<string, unknown>;
    const role = obj['role'];
    if (typeof role === 'string' && role.length > 0) {
      return ROLE_ALIASES[role] ?? role;
    }
    const raw = String(obj['role'] || obj['type'] || '');
    if (raw.length > 0) return ROLE_ALIASES[raw] ?? raw;
  }
  // 类名兜底：仅对自定义类生效，构造来源为 Object/无自定义类的纯字典不回退
  if (isObj) {
    const ctor = (msg as { constructor?: { name?: string } }).constructor;
    const ctorName = ctor?.name ?? '';
    if (ctorName && ctorName !== 'Object') {
      const name = ctorName.toLowerCase();
      const stem = name.endsWith('message') ? name.slice(0, -'message'.length) : name;
      return ROLE_ALIASES[stem] ?? (stem || name);
    }
  }
  return '';
}

/** 按 index 合并流式工具调用增量（保持首次出现顺序，index 乱序时仍按首见序输出）。 */
export function accumulate_tool_calls(deltas: readonly ToolCallDelta[]): ToolCall[] {
  const byIndex = new Map<number, ToolCall>();
  const order: number[] = [];
  for (const delta of deltas) {
    let tc = byIndex.get(delta.index);
    if (tc === undefined) {
      tc = new ToolCall({ id: delta.id ?? '', name: delta.name ?? '', arguments: '' });
      byIndex.set(delta.index, tc);
      order.push(delta.index);
    }
    if (delta.id) tc.id = delta.id;
    if (delta.name) tc.name = delta.name;
    if (delta.arguments_delta) tc.arguments += delta.arguments_delta;
  }
  return order.map((i) => byIndex.get(i) as ToolCall);
}